crates/storage/provider/src/providers/mod.rs

```diff
@@ -94,10 +94,20 @@ impl<DB: Database> HeaderProvider for ShareableDatabase<DB> {
 }
 
 impl<DB: Database> BlockHashProvider for ShareableDatabase<DB> {
-    fn block_hash(&self, number: U256) -> Result<Option<H256>> {
-        // TODO: This unwrap is potentially unsafe
+    fn block_hash(&self, number: u64) -> Result<Option<H256>> {
+        self.db.view(|tx| tx.get::<tables::CanonicalHeaders>(number))?.map_err(Into::into)
+    }
+
+    fn canonical_hashes_range(&self, start: BlockNumber, end: BlockNumber) -> Result<Vec<H256>> {
+        let range = start..end;
         self.db
-            .view(|tx| tx.get::<tables::CanonicalHeaders>(number.try_into().unwrap()))?
+            .view(|tx| {
+                let mut cursor = tx.cursor_read::<tables::CanonicalHeaders>()?;
+                cursor
+                    .walk_range(range)?
+                    .map(|result| result.map(|(_, hash)| hash).map_err(Into::into))
+                    .collect::<Result<Vec<_>>>()
+            })?
             .map_err(Into::into)
     }
 }
```

Comment:
@rakita @onbjerg mentioned that we might be doing redundant reads per block to get the last 256 blockhashes, probably related to this. it'd be nice if we could cache the last 256 blocks in memory, and when new blcok arrives only read for that, instead of walking many blocks. maybe not a big perf gain.